src/core/src/pass/serialize.cpp

```diff
@@ -105,56 +105,50 @@ class ConstantWriter {
+                auto fp16_buffer = compress_data_to_fp16(ptr, size, src_type, new_size);
+                m_binary_output.write(fp16_buffer.get(), *new_size);
+            }
             return offset;
-        }
-        // TODO: Find a way to keep both types of compression (m_enable_compression and compress_to_fp16)
-        // simultaneously. Disabled usual compression by m_enable_compression for those constants that are requested to
-        // be compressed by compress_to_fp16 for now. To implement both compression types applied simultaneously
-        // we need to save element_type for each constant in the cache together with the compression status
-        // that implies a wider impact and requires a more accurate implementation of cache handling.
-        // When FP16 compression is turned on together with the usual compression enabled by m_enable_compression, we
-        // can avoid comparing FP32 weights, but it would require comparing with data from a file, because on-the-fly
-        // converted FP16 constants are not kept in memory.
-
-        // This hash is weak (but efficient) and must be replace with some other
-        // more stable hash algorithm. For example current hash algorithms gives
-        // the same hash for {2, 2} and {0, 128} arrays. So we have to compare
-        // values when finding a match in hash map.
-        const HashValue hash = hash_combine(ptr, size);
-        const auto found = m_hash_to_file_positions.find(hash);
-        if (found != end(m_hash_to_file_positions) &&
-            memcmp(static_cast<void const*>(ptr), found->second.second, size) == 0) {
-            return found->second.first;
-        }
-
-        write_with_optional_fp16_compression(ptr, size, new_size, compress_to_fp16, src_type);
-        m_hash_to_file_positions.insert({hash, {offset, static_cast<void const*>(ptr)}});
+        } else {
+            std::unique_ptr<char[]> fp16_buffer = nullptr;
+            if (compress_to_fp16) {
+                OPENVINO_ASSERT(size % src_type.size() == 0);
+                fp16_buffer = compress_data_to_fp16(ptr, size, src_type, new_size);
+            }
+            const char* ptr_to_write;
+            if (fp16_buffer) {
+                ptr_to_write = fp16_buffer.get();
+            } else {
+                ptr_to_write = ptr;
+            }
 
+            // This hash is weak (but efficient). For example current hash algorithms gives
+            // the same hash for {2, 2} and {0, 128} arrays.
+            // But even strong hashing algorithms sometimes give collisions.
+            // Therefore we always have to compare values when finding a match in hash map.
+            const HashValue hash = hash_combine(ptr_to_write, *new_size);
+            const auto found = m_hash_to_file_positions.find(hash);
+            if (found != end(m_hash_to_file_positions) && memcmp(ptr, found->second.second, size) == 0) {
+                return found->second.first;
+            }
```

Comment:
What if we have several collisions and one of them with the same value? Here we just check only one collision and forget about others that can be with the same value

Also, can we have a tests with collision cases?

Reply:
That's a good question! I thought about that.

If we have 2 or more constants with the same hashes but different values, only the last of them will be stored in the map since keys of `m_hash_to_file_positions` are unique. This means that if e.g. we further see the constant that equals to the first one instead of referring to the existing place in the file we will again write it and rewrite value for that hash in the map. For several collision the same will happen, only the last value will be stored and could be 'compressed'. To overcome that we would need to replace `unordered_map` to `multimap` in 
https://github.com/openvinotoolkit/openvino/blob/7f5563e6f0ff7dbc04cd864e4062282c06943b5a/src/core/src/pass/serialize.cpp#L92

About tests for collisions: there is already a test for that https://github.com/openvinotoolkit/openvino/blob/7f5563e6f0ff7dbc04cd864e4062282c06943b5a/src/core/tests/pass/serialization/const_compression.cpp#L126-L142

I wanted to add a couple more, but could find more collision myself, they are very very rare, i suspect multiple collisions are even more rare. That was the reason I stopped in the existing solution instead of replacing map to multimap. But if you say, i can do with multimap

Reply:
Replaced with `multimap`. If there are double collision it will store both of them and refer to the same place if we see constant which is equal to any of them, not just if it equal only to the last of them. Added unit-test for that.

Same will happen if we face triple and more collisions, we store them all and `while (found != m_hash_to_file_positions.end())` will iterate through all of them. It's a very very rare event, unfortunately i didn't find values which collide triple or more time to add such unit-test, but with multimap test above can guarantee that this will work well. @rkazants could please take a look to the latest solution?